File: 124_binary_tree_maximum_path_sum/124_binary_tree_maximum_path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long ans;
    int maxPathSum(TreeNode* root) {
        ans = root->val;
        maxSumHelper(root)
        return ans;
    }
    int maxSumHelper(TreeNode* root) {
        if(root == nullptr) return 0;
        int leftPathSum = max(0, maxSumHelper(root->left));
        int rightPathSum = max(0, maxSumHelper(root->right, res));
        ans = max(ans, leftPathSum + rightPathSum + root->val);
        return max(leftPathSum, rightPathSum) + root->val; // so return the maximum path rooted at this root
    }
};

File: 124. Binary Tree Maximum Path Sum/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import Optional
from collections import defaultdict, deque
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        ans = root.val
        l = defaultdict(int)
        r = defaultdict(int)
        
        def dfs(current: TreeNode):
            nonlocal ans
            if current is None: return 0
            left, right = current.left, current.right
            leftVal = max(0, dfs(left))
            rightVal = max(0, dfs(right))
            ans = max(ans, leftVal + current.val + rightVal)
            return current.val + max(leftVal, rightVal)
        dfs(root)
        return ans
    
root = TreeNode(-10)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)
sol = Solution()
print(sol.maxPathSum(root))